README.md
![alt tag](https://s24.postimg.org/ok064585x/dotted-progress-title.png)

[![CI Status](http://img.shields.io/travis/nikola9core/DottedProgressBar.svg?style=flat)](https://travis-ci.org/nikola9core/DottedProgressBar)
[![Version](https://img.shields.io/cocoapods/v/DottedProgressBar.svg?style=flat)](http://cocoapods.org/pods/DottedProgressBar)
[![License](https://img.shields.io/cocoapods/l/DottedProgressBar.svg?style=flat)](http://cocoapods.org/pods/DottedProgressBar)
[![Platform](https://img.shields.io/cocoapods/p/DottedProgressBar.svg?style=flat)](http://cocoapods.org/pods/DottedProgressBar)

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

![alt tag](https://gifyu.com/images/dotted-progress-bar-ezgif-480.gif)

## Requirements
* iOS 8.0+
* Swift 3.0+

## Installation

DottedProgressBar is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "DottedProgressBar"
```
## Easy to use
Import library
```swift
import DottedProgressBar
```

Initialize `DottedProgressBar` in one line of code
```swift
let progressBar = DottedProgressBar(frame: CGRect(x: 50, y: 50, width: 200, height: 20),
                                    numberOfDots: 6,
                                    initialProgress: 1)
view.addSubview(progressBar)
```

## Custom appearance
```swift
let progressBar = DottedProgressBar()
progressBar.appearance = DottedProgressBar.DottedProgressAppearance(
    dotRadius: 8.0,
    dotsColor: UIColor.orange.withAlphaComponent(0.5),
    dotsProgressColor: UIColor.red,
    backColor: UIColor.clear
)
view.addSubview(progressBar)
progressBar.frame = CGRect(x: 50, y: 50, width: 200, height: 20)

progressBar.setNumberOfDots(6, animated: false)
progressBar.setProgress(1, animated: false)
```

## Animations
Animations can be called repeatedly because they have theirs own queue. Each animation will wait previous to finish and then will be executed.
```swift
self.setProgress(4, animated: true)
self.setNumberOfDots(8, animated: true)
```

## Customize animations

Customize duration of animations and pause between consecutive animations
```swift
progressBar.dotsNumberChangeAnimationDuration = 0.6
progressBar.progressChangeAnimationDuration = 0.7
progressBar.pauseBetweenConsecutiveAnimations = 1.0
```

## License

DottedProgressBar is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Xcodeproj Changelog

## 1.0.4

##### General

* Swift 4 support



Example/Tests/Tests.swift
//// https://github.com/Quick/Quick
//
//import Quick
//import Nimble
//import DottedProgressBar
//import UIKit
//

import XCTest
@testable import DottedProgressBar
class DottedProgressBarAppTests: XCTestCase {
    
    let progressBar = DottedProgressBar()
    
    override func setUp() {
        super.setUp()
        
        //appearance
        progressBar.progressAppearance = DottedProgressBar.DottedProgressAppearance(
            dotRadius: 8.0,
            dotsColor: UIColor.orange.withAlphaComponent(0.4),
            dotsProgressColor: UIColor.red,
            backColor: UIColor.clear
        )
        
        //set number of steps and current progress
        progressBar.setNumberOfDots(6, animated: false)
        progressBar.setProgress(1, animated: false)
        
        //customize animation
        progressBar.dotsNumberChangeAnimationDuration = 0.6
        progressBar.progressChangeAnimationDuration = 0.7
        progressBar.pauseBetweenConsecutiveAnimations = 1.0
        
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testProgressBarSubviews() {
        let numberOfSubviews = progressBar.subviews.count
        XCTAssertTrue(numberOfSubviews == 6)
    }
    
}


//class TableOfContentsSpec: QuickSpec {
//    override func spec() {
//        describe("dotted progress bar view") {
//
//            it("count dots after async animation") {
//                let progressBar = DottedProgressBar()
//                progressBar.setNumberOfDots(6, animated: true)
//
//                //toEventually will check condition after animation completed
//                expect(progressBar.subviews.count).toEventually(equal(6))
//            }
//
//            it("checks dots color after async animation") {
//                waitUntil { done in
//                    let progressBar = DottedProgressBar()
//                    progressBar.progressAppearance = DottedProgressBar.DottedProgressAppearance(
//                        dotRadius: 8.0,
//                        dotsColor: UIColor.orange,
//                        dotsProgressColor: UIColor.red,
//                        backColor: UIColor.clear
//                    )
//
//                    progressBar.setNumberOfDots(8, animated: false)
//                    progressBar.setProgress(3, animated: true)
//
//                    DottedBarUtility.delay(progressBar.progressChangeAnimationDuration + 0.1) {
//                        expect(progressBar.subviews[2].backgroundColor).to(equal(UIColor.red))
//                        expect(progressBar.subviews[3].backgroundColor).to(equal(UIColor.orange))
//                        done()
//                    }
//                }
//            }
//
//        }
//    }
//}



Example/DottedProgressBar/AppDelegate.swift
//
//  AppDelegate.swift
//  DottedProgressBar
//
//  Created by nikola9core on 04/16/2017.
//  Copyright (c) 2017 nikola9core. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    var window: UIWindow?
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }
    
    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }
    
    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }
    
    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }
    
    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }
    
}



DottedProgressBar/Classes/DottedProgressBar.swift
//
//  DottedProgressBar.swift
//  Pods
//
//  Created by Nikola Corlija on 4/16/17.
//
//

import Foundation
import UIKit

// swiftlint:disable function_body_length

open class DottedProgressBar: UIView {

    public struct DottedProgressAppearance {
        let dotRadius: CGFloat
        let dotsColor: UIColor
        let dotsProgressColor: UIColor
        let backColor: UIColor

        public init(dotRadius: CGFloat = 8.0,
                    dotsColor: UIColor = UIColor.orange.withAlphaComponent(0.4),
                    dotsProgressColor: UIColor = UIColor.red,
                    backColor: UIColor = UIColor.clear) {
            self.dotRadius = dotRadius
            self.dotsColor = dotsColor
            self.dotsProgressColor = dotsProgressColor
            self.backColor = backColor
        }
    }

    open var progressAppearance: DottedProgressAppearance!

    /// The duration of dots number change animation in seconds.
    open var dotsNumberChangeAnimationDuration: Double = 0.7
    /// The duration of dots progress change animation in seconds.
    open var progressChangeAnimationDuration: Double = 0.8
    /// The time between performing animations from animation queue in seconds.
    open var pauseBetweenConsecutiveAnimations: Double = 1.0
    /// Zoom increase of walking dot while animating progress.
    open var zoomIncreaseValueOnProgressAnimation: CGFloat = 1.5

    fileprivate var numberOfDots: Int = 1
    fileprivate var previousProgress: Int = 0
    fileprivate var currentProgress: Int = 0

    fileprivate lazy var animationQueue = DottedBarAnimationQueue()
    fileprivate var isAnimatingCurrently: Bool = false
    fileprivate lazy var walkingDot = UIView()

    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override public init(frame: CGRect) {
        progressAppearance = DottedProgressAppearance()
        super.init(frame: frame)
        setup()
    }

    public init(frame: CGRect, numberOfDots: Int, initialProgress: Int) {
        progressAppearance = DottedProgressAppearance()
        super.init(frame: frame)
        self.numberOfDots = numberOfDots
        self.currentProgress = initialProgress
        setup()
    }

    public init(appearance: DottedProgressAppearance) {
        self.progressAppearance = appearance
        super.init(frame: CGRect.zero)
        setup()
    }

    /// Sets a number of steps of progress bar with or without animation.
    ///
    /// - Parameters:
    ///   - count: Number of steps/dots.
    ///   - animated: Flag for animate effect.
    open func setNumberOfDots(_ count: Int, animated: Bool = true) {
        animationQueue.enqueue(DottedBarAnimation(type: .numberChange, value: count, animated: animated))
        if !isAnimatingCurrently {
            performQueuedAnimations()
        }
    }

    /// Sets a number of filled dots as current progress with or without animation.
    ///
    /// - Parameters:
    ///   - progress: Number of steps/dots of current progress.
    ///   - animated: Flag for animate effect.
    open func setProgress(_ progress: Int, animated: Bool = true) {
        animationQueue.enqueue(DottedBarAnimation(type: .progresChange, value: progress, animated: animated))
        if !isAnimatingCurrently {
            performQueuedAnimations()
        }
    }

    override open func layoutSubviews() {
        super.layoutSubviews()
        layout()
    }

}

// MARK: - Private functions

private extension DottedProgressBar {

    func setup() {
        backgroundColor = progressAppearance.backColor

        for i in 0..<numberOfDots {
            let dot = UIView()
            dot.backgroundColor = i < currentProgress ? progressAppearance.dotsProgressColor :
                progressAppearance.dotsColor
            dot.layer.cornerRadius = progressAppearance.dotRadius
            dot.frame = dotFrame(forIndex: i)
            addSubview(dot)
        }
    }

    func layout() {
        for (index, dot) in subviews.enumerated() where dot != walkingDot {
            dot.layer.cornerRadius = progressAppearance.dotRadius
            dot.frame = dotFrame(forIndex: index)
        }
    }

    /// Calculating frame for given index of dot, supports vertical and horizontal alignment.
    ///
    /// - Parameter index: Index of dot (including 0).
    /// - Returns: Frame rectangle for given dot index
    func dotFrame(forIndex index: Int) -> CGRect {
        guard index >= 0 else {
            return dotFrame(forIndex: 0)
        }
        if frame.size.width > frame.size.height {
            let externalFrameWidth: CGFloat = frame.size.width / CGFloat(numberOfDots)
            let externalFrame = CGRect(x: CGFloat(index) * externalFrameWidth,
                                       y: 0, width: externalFrameWidth,
                                       height: frame.size.height)
            return CGRect(x: externalFrame.midX - progressAppearance.dotRadius,
                          y: externalFrame.midY - progressAppearance.dotRadius,
                          width: progressAppearance.dotRadius * 2,
                          height: progressAppearance.dotRadius * 2)
        } else {
            let externalFrameHeight: CGFloat = frame.size.height / CGFloat(numberOfDots)
            let externalFrame = CGRect(x: 0,
                                       y: CGFloat(index) * externalFrameHeight,
                                       width: frame.size.width,
                                       height: externalFrameHeight)
            return CGRect(x: externalFrame.midX - progressAppearance.dotRadius,
                          y: externalFrame.midY - progressAppearance.dotRadius,
                          width: progressAppearance.dotRadius * 2,
                          height: progressAppearance.dotRadius * 2)
        }
    }

    /// Starting execution of all queued animations.
    func performQueuedAnimations() {
        if let nextAnimation = animationQueue.dequeue() {
            isAnimatingCurrently = true
            if nextAnimation.type == .numberChange {
                if nextAnimation.value > 0 &&
                    nextAnimation.value >= currentProgress &&
                    nextAnimation.value != numberOfDots {
                    animateNumberChange(animation: nextAnimation)
                } else {
                    print("DottedProgressBar - invalid setNumberOfDots \(nextAnimation.value)")
                    self.performQueuedAnimations()
                }
            } else {
                if nextAnimation.value > 0 &&
                    nextAnimation.value <= numberOfDots &&
                    nextAnimation.value != currentProgress {
                    animateProgress(animation: nextAnimation)
                } else {
                    print("DottedProgressBar - invalid setProgress \(nextAnimation.value)")
                    self.performQueuedAnimations()
                }
            }
        } else {
            isAnimatingCurrently = false
        }
    }

    /// Performs animation for changing the number of dots
    ///
    /// - Parameter animation: The animation model
    func animateNumberChange(animation: DottedBarAnimation) {
        numberOfDots = animation.value

        if numberOfDots > subviews.count {
            UIView.animate(withDuration: animation.animated ? dotsNumberChangeAnimationDuration * 0.6 : 0.0,
                           delay: 0.0, options: UIViewAnimationOptions(), animations: {
                            self.layout()
            }, completion: { _ in
                for _ in 0 ..< (self.numberOfDots - self.subviews.count) {
                    let view = UIView()
                    view.backgroundColor = self.progressAppearance.dotsColor
                    view.layer.cornerRadius = self.progressAppearance.dotRadius
                    view.alpha = 0
                    self.addSubview(view)
                }
                UIView.animate(withDuration: animation.animated ? self.dotsNumberChangeAnimationDuration * 0.4 : 0.0,
                               delay: 0.0,
                               options: UIViewAnimationOptions(), animations: {
                                for dot in self.subviews {
                                    dot.alpha = 1
                                }
                }, completion: { _ in
                    if animation.animated {
                        DottedBarUtility.delay(self.pauseBetweenConsecutiveAnimations, closure: {
                            self.performQueuedAnimations()
                        })
                    } else {
                        self.performQueuedAnimations()
                    }
                })

                self.layout()
            })
        } else {
            UIView.animate(withDuration: animation.animated ? self.dotsNumberChangeAnimationDuration * 0.4 : 0.0,
                           delay: 0.0,
                           options: UIViewAnimationOptions(),
                           animations: {
                            for index in (Int(self.numberOfDots)..<self.subviews.count).reversed() {
                                self.subviews[index].alpha = 0
                            }
            }, completion: { _ in
                for index in (Int(self.numberOfDots)..<self.subviews.count).reversed() {
                    self.subviews[index].removeFromSuperview()
                }
                UIView.animate(withDuration: animation.animated ? self.dotsNumberChangeAnimationDuration * 0.6 : 0.0,
                               delay: 0.0,
                               options: UIViewAnimationOptions(),
                               animations: {
                                self.layout()
                }, completion: { _ in
                    if animation.animated {
                        DottedBarUtility.delay(self.pauseBetweenConsecutiveAnimations, closure: {
                            self.performQueuedAnimations()
                        })
                    } else {
                        self.performQueuedAnimations()
                    }
                })
            })
        }
    }

    /// Performs animation for changing the current progress
    ///
    /// - Parameter animation: The animation model
    func animateProgress(animation: DottedBarAnimation) {
        previousProgress = currentProgress
        currentProgress = animation.value

        if animation.animated {
            walkingDot.backgroundColor = progressAppearance.dotsProgressColor
            walkingDot.layer.cornerRadius = progressAppearance.dotRadius
            walkingDot.frame = dotFrame(forIndex: previousProgress - 1)
            addSubview(walkingDot)
            walkingDot.layer.zPosition = 1

            UIView.animate(withDuration: progressChangeAnimationDuration * 0.7,
                           delay: 0.0,
                           options: .curveLinear,
                           animations: {
                            let frame = self.dotFrame(forIndex: self.currentProgress - 1)
                            self.walkingDot.frame = CGRect(x: frame.origin.x - self.zoomIncreaseValueOnProgressAnimation,
                                                           y: frame.origin.y - self.zoomIncreaseValueOnProgressAnimation,
                                                           width:
                                self.progressAppearance.dotRadius * 2 +
                                    self.zoomIncreaseValueOnProgressAnimation * 2,
                                                           height:
                                self.progressAppearance.dotRadius * 2 +
                                    self.zoomIncreaseValueOnProgressAnimation * 2)
                            self.walkingDot.layer.cornerRadius =
                                self.progressAppearance.dotRadius +
                                self.zoomIncreaseValueOnProgressAnimation
            }, completion: nil)
        }

        let dotsRange: CountableClosedRange = currentProgress > previousProgress ?
            previousProgress...currentProgress - 1 :
            currentProgress...previousProgress - 1

        for index in dotsRange {
            UIView.animate(withDuration: 0.1,
                           delay:
                animation.animated ?
                    progressChangeAnimationDuration * 0.7 *
                        ((Double(index) - Double(previousProgress - 1)) / (Double(currentProgress - 1) -
                            Double(previousProgress - 1))) : 0.0,
                           options: .curveLinear,
                           animations: {
                            self.subviews[index].backgroundColor =
                                self.currentProgress > self.previousProgress ?
                                    self.progressAppearance.dotsProgressColor :
                                self.progressAppearance.dotsColor
            }, completion: nil)
        }

        if animation.animated {
            UIView.animate(withDuration: progressChangeAnimationDuration * 0.3,
                           delay: progressChangeAnimationDuration * 0.7,
                           options: UIViewAnimationOptions(),
                           animations: {
                            self.walkingDot.frame = self.dotFrame(forIndex: self.currentProgress - 1)
            }, completion: { _ in
                self.walkingDot.removeFromSuperview()
                DottedBarUtility.delay(self.pauseBetweenConsecutiveAnimations, closure: {
                    self.performQueuedAnimations()
                })
            })
        } else {
            self.performQueuedAnimations()
        }
    }
}


DottedProgressBar/Classes/DottedBarAnimationQueue.swift
//
//  DottedBarAnimationQueue.swift
//  Pods
//
//  Created by Nikola Corlija on 4/16/17.
//
//

import Foundation

struct DottedBarAnimationQueue {
    
    var queue = [DottedBarAnimation]()
    
    var isEmpty: Bool {
        return queue.isEmpty
    }
    
    mutating func enqueue(_ element: DottedBarAnimation) {
        queue.append(element)
    }
    
    mutating func dequeue() -> DottedBarAnimation? {
        if !queue.isEmpty {
            return queue.removeFirst()
        } else {
            return nil
        }
    }
    
    func peek() -> DottedBarAnimation? {
        if !queue.isEmpty {
            return queue[0]
        } else {
            return nil
        }
    }
    
}

struct DottedBarAnimation {
    var type: DottedBarAnimationType = .numberChange
    var value: Int = 0
    var animated: Bool = false
}

/// Animation type while setting number of dots or current progress
///
/// - numberChange: Change of number of progress steps
/// - progresChange: Change of progress
enum DottedBarAnimationType {
    case numberChange
    case progresChange
}

public class DottedBarUtility {
    /// Utility static function for delaying job on main thread.
    ///
    /// - Parameters:
    ///   - delay: The time interval for delaying in seconds.
    ///   - closure: The code to be executed after delay time passes.
    public static func delay(_ delay: Double, closure: @escaping () -> ()) {
        DispatchQueue.main.asyncAfter(
            deadline: DispatchTime.now() + delay, execute: closure
        )
    }
}


Example/DottedProgressBar/ViewController.swift
//
//  ViewController.swift
//  DottedProgressBar
//
//  Created by nikola9core on 04/16/2017.
//  Copyright (c) 2017 nikola9core. All rights reserved.
//

import UIKit
import DottedProgressBar

class ViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let progressBar = DottedProgressBar()
        
        //appearance
        progressBar.progressAppearance = DottedProgressBar.DottedProgressAppearance(
            dotRadius: 8.0,
            dotsColor: UIColor.orange.withAlphaComponent(0.4),
            dotsProgressColor: UIColor.red,
            backColor: UIColor.clear
        )
        
        progressBar.frame = CGRect(x: 50, y: 50, width: 200, height: 20)
        view.addSubview(progressBar)
        
        //set number of steps and current progress
        progressBar.setNumberOfDots(6, animated: false)
        progressBar.setProgress(1, animated: false)
        
        //customize animation
        progressBar.dotsNumberChangeAnimationDuration = 0.6
        progressBar.progressChangeAnimationDuration = 0.7
        progressBar.pauseBetweenConsecutiveAnimations = 1.0
        
        //perform animations
        progressBar.setProgress(4)
        progressBar.setNumberOfDots(7)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
}

